### PR TITLE
merge dev-alpha into dev.

### PR DESCRIPTION
**Pull request by devinleighsmith**

### CODE REVIEW

File: frontend/src/components/common/form/ParentSelect.tsx [hunk 1]
```diff
@@ -7,6 +7,7 @@ import { Highlighter, Menu, MenuItem } from 'react-bootstrap-typeahead';
 import { Label } from '../Label';
```
Comment (devinleighsmith) [inline frontend/src/components/common/form/ParentSelect.tsx:7]: @jdtoombs please review these changes, I've updated the components that use ParentSelect to handle the change from returning an object to returning an id by default.

File: backend/api/Areas/Property/Models/Building/LeasedLandMetadataModel.cs [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+using System;
+using System.Collections.Generic;
+
+namespace Pims.Api.Areas.Property.Models.Building
+{
+    public class LeasedLandMetadataModel
+    {
+        #region Properties
+        /// <summary>
+        /// get/set - Whether an exemption was requested for the ERP.
```
Comment (Fosol) [inline backend/api/Areas/Property/Models/Building/LeasedLandMetadataModel.cs:12]: Doc

File: backend/api/Models/Parcel/ParcelModel.cs [hunk 3]
```diff
@@ -21,7 +21,7 @@ public class ParcelModel : PropertyModel
 
         public IEnumerable<ParcelFiscalModel> Fiscals { get; set; } = new List<ParcelFiscalModel>();
 
-        public IEnumerable<ParcelBuildingModel> Buildings { get; set; } = new List<ParcelBuildingModel>();
+        public virtual IEnumerable<ParcelBuildingModel> Buildings { get; set; } = new List<ParcelBuildingModel>();
```
Comment (Fosol) [inline backend/api/Models/Parcel/ParcelModel.cs:24]: Is there a reason it's `virtual`?

File: backend/dal/Helpers/Extensions/BuildingExtensions.cs [hunk 4]
```diff
@@ -24,7 +24,11 @@ public static class BuildingExtensions
         /// <exception type="DbUpdateException">The name within a parcel should be unique.</exception>
         public static void ThrowIfNotUnique(this PimsContext context, Entity.Parcel parcel, Entity.Building building)
         {
-            var parcelBuildings = context.Parcels.Where(p => p.Id == parcel.Id).SelectMany(p => p.Buildings.Select(b => b.Building.Name)).Distinct().ToArray();
+            if(parcel == null || String.IsNullOrEmpty(building.Name))
```
Comment (Fosol) [inline backend/dal/Helpers/Extensions/BuildingExtensions.cs:27]: If it's not named or belongs to a parcel don't worry about being unique?

File: backend/dal/Migrations/20201217172215_v01.05.03.Designer.cs [hunk 5]
```diff
@@ -0,0 +1,3415 @@
+﻿// <auto-generated />
```
Comment (Fosol) [inline backend/dal/Migrations/20201217172215_v01.05.03.Designer.cs:1]: I think this file should be deleted.

File: backend/dal/Migrations/20201217172215_v01.05.03.cs [hunk 6]
```diff
@@ -0,0 +1,28 @@
+﻿using Microsoft.EntityFrameworkCore.Migrations;
```
Comment (Fosol) [inline backend/dal/Migrations/20201217172215_v01.05.03.cs:1]: This migration needs to be merged into `v01.06.00` and then deleted.

File: backend/dal/Migrations/20210106182851_v01.06.01.cs [hunk 7]
```diff
@@ -0,0 +1,47 @@
+﻿using System;
```
Comment (Fosol) [inline backend/dal/Migrations/20210106182851_v01.06.01.cs:1]: As you'll need to update `v01.06.00` from merging `v01.05.03` you may as well merge this back into `v01.06.00` too

File: backend/dal/Services/Concrete/BuildingService.cs [hunk 8]
```diff
@@ -193,14 +227,132 @@ public Building Update(Building building)
...
+            }
+
+            if (allowEdit)
+            {
+                this.Context.Entry(existingBuilding.Address).CurrentValues.SetValues(building.Address);
+                this.Context.Entry(existingBuilding).CurrentValues.SetValues(building);
+                this.Context.SetOriginalRowVersion(existingBuilding);
+            }
+
+            if (allowEdit)
```
Comment (Fosol) [inline backend/dal/Services/Concrete/BuildingService.cs:270]: Any reason you can't include this if statement with the one above?

File: backend/dal/Services/Concrete/BuildingService.cs [hunk 9]
```diff
@@ -193,14 +227,132 @@ public Building Update(Building building)
...
+                    // Delete the fiscals from the parcel that have been removed.
+                    if (!building.Fiscals.Any(e => (e.BuildingId == buildingFiscals.BuildingId && e.FiscalYear == buildingFiscals.FiscalYear && e.Key == buildingFiscals.Key)))
+                    {
+                        this.Context.BuildingFiscals.Remove(buildingFiscals);
+                    }
+                }
+            }
+            existingBuilding.LeasedLandMetadata = building.LeasedLandMetadata;
             this.Context.Buildings.Update(existingBuilding); // TODO: Must detach entity before returning it.
             this.Context.CommitTransaction();
```
Comment (Fosol) [inline backend/dal/Services/Concrete/BuildingService.cs:354]: Have you tested for concurrency with `RowVersion`?

File: backend/dal/Services/Concrete/ParcelService.cs [hunk 10]
```diff
@@ -212,6 +213,20 @@ public Parcel Add(Parcel parcel)
...
+            return parcel;
+        }
+
+        /// <summary>
+        /// Update the specified parcel in the datasource, but do not commit the transaction.
+        /// </summary>
+        /// <param name="parcel"></param>
+        /// <exception type="KeyNotFoundException">Entity does not exist in the datasource.</exception>
+        /// <returns></returns>
+        public Parcel PendingUpdate(Parcel parcel)
```
Comment (Fosol) [inline backend/dal/Services/Concrete/ParcelService.cs:229]: I like this naming convention `Pending*`.  We should use it more.

File: backend/entities/BuildingFiscal.cs [hunk 11]
```diff
@@ -26,6 +26,12 @@ public class BuildingFiscal : BaseEntity
         /// <value></value>
         public int FiscalYear { get; set; }
 
+        /// <summary>
+        /// get/set - The effective date of this fiscal value
+        /// </summary>
+        /// <value></value>
+        public DateTime EffectiveDate { get; set; }
```
Comment (Fosol) [inline backend/entities/BuildingFiscal.cs:33]: As this isn't nullable, how will existing data be handled in the migration?

File: backend/entities/Models/LeasedLandMetadata.cs [hunk 12]
```diff
@@ -0,0 +1,22 @@
+using System;
+
+namespace Pims.Dal.Entities.Models
+{
+    public class LeasedLandMetadata
+    {
+        #region Properties
+        /// <summary>
+        /// get/set - Whether an exemption was requested for the ERP.
```
Comment (Fosol) [inline backend/entities/Models/LeasedLandMetadata.cs:9]: doc

File: backend/entities/Models/LeasedLandMetadata.cs [hunk 13]
```diff
@@ -0,0 +1,22 @@
...
+        /// </summary>
+        public string OwnershipNote { get; set; }
+        /// <summary>
+        /// get/set - An id used to associate this metadata with a parcel.
+        /// </summary>
+        public string ParcelId { get; set; }
+        /// <summary>
+        /// get/set - the lease type
+        /// </summary>
+        public int Type { get; set; }
```
Comment (Fosol) [inline backend/entities/Models/LeasedLandMetadata.cs:19]: I suspect we should change the `int` to an `enum`

File: backend/entities/ParcelFiscal.cs [hunk 14]
```diff
@@ -26,6 +26,12 @@ public class ParcelFiscal : BaseEntity
         /// <value></value>
         public int FiscalYear { get; set; }
 
+        /// <summary>
+        /// get/set - The effective date of this fiscal value
+        /// </summary>
+        /// <value></value>
+        public DateTime? EffectiveDate { get; set; }
```
Comment (Fosol) [inline backend/entities/ParcelFiscal.cs:33]: `BuildingFiscal` is not nullable, but here a `ParselFiscal` is?  Which is correct?

File: frontend/package.json [hunk 15]
```diff
@@ -124,7 +126,7 @@
     "test": "react-scripts test --env=jsdom-fourteen --ci --watchAll=false --runInBand",
     "test:watch": "react-scripts test --env=jsdom-fourteen",
     "coverage": "react-scripts test --ci --coverage --testResultsProcessor jest-sonar-reporter --env=jsdom-fourteen --watchAll=false --runInBand",
-    "update-snapshots": "react-scripts test -- --updateSnapshot",
+    "update-snapshots": "react-scripts test --env=jsdom-fourteen --updateSnapshot",
```
Comment (Fosol) [inline frontend/package.json:129]: What's the `env` for?

File: frontend/src/actions/parcelsActions.ts [hunk 16]
```diff
@@ -73,16 +80,52 @@ export interface IBuilding extends IProperty {
...
   buildingTenancy: string;
+  buildingTenancyUpdatedOn?: string;
   rentableArea: number | '';
+  totalArea: number | '';
   agencyId: number | '';
   agency: string;
   agencyCode: string;
   evaluations: IEvaluation[];
   fiscals: IFiscal[];
+  leasedLand: {
```
Comment (Fosol) [inline frontend/src/actions/parcelsActions.ts:96]: Should there be an interface for `ILeasedLand`?

**Review by jdtoombs — APPROVED** [review]
Looks good re the parent select and typeahead!

File: backend/api/Models/Parcel/ParcelModel.cs (hunk 3 above)
Comment (devinleighsmith) [reply backend/api/Models/Parcel/ParcelModel.cs:24]: I'll check, I thought I was overriding this somewhere.

File: backend/dal/Helpers/Extensions/BuildingExtensions.cs (hunk 4 above)
Comment (devinleighsmith) [reply backend/dal/Helpers/Extensions/BuildingExtensions.cs:27]: Yes, unfortunately we need this as there are parcels with buildings that already violate this rule within the DB. For example
Parcel 1
  Building 1 Name: ''
  Building 2 Name: ''
  Building 3 Name: 'Test'
If I modify Building 3, this check was previously throwing an exception.

File: backend/dal/Services/Concrete/BuildingService.cs (hunk 8 above)
Comment (devinleighsmith) [reply backend/dal/Services/Concrete/BuildingService.cs:270]: Not that I can see.

File: backend/dal/Services/Concrete/BuildingService.cs (hunk 9 above)
Comment (devinleighsmith) [reply backend/dal/Services/Concrete/BuildingService.cs:354]: Yes

File: backend/entities/BuildingFiscal.cs (hunk 11 above)
Comment (devinleighsmith) [reply backend/entities/BuildingFiscal.cs:33]: Existing rows would be assigned 01/01/0001, so will probably need to make nullable

File: frontend/package.json (hunk 15 above)
Comment (devinleighsmith) [reply frontend/package.json:129]: try running the old version, it won't work. the jsdom-fourteen has some required features we need to make the snapshots work.

File: backend/entities/ParcelFiscal.cs (hunk 14 above)
Comment (devinleighsmith) [reply backend/entities/ParcelFiscal.cs:33]: they should both be nullable.

File: backend/api/Models/Parcel/ParcelModel.cs (hunk 3 above)
Comment (devinleighsmith) [reply backend/api/Models/Parcel/ParcelModel.cs:24]: Yes, overriden in the buildingParcelModel

**Review by agalbrai — COMMENTED** [review]
A round of applause for you sir, this must have been a huge amount of work. Fingers crossed it goes over well

**Review by Fosol — APPROVED** [review]

**Pull request merged**